torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -28,12 +89,54 @@ class SubgraphSlicer {
   SubgraphSlicer(
       Block* block,
       std::shared_ptr<Graph> graph,
-      size_t minSubgraphSize)
+      size_t minSubgraphSize,
+      AliasDb& aliasDb,
+      std::vector<Node*>& diff_nodes)
       : block_(block),
         graph_(std::move(graph)),
-        minSubgraphSize_(minSubgraphSize) {}
+        minSubgraphSize_(minSubgraphSize),
+        aliasDb_(aliasDb),
+        diff_nodes_(diff_nodes) {}
+
+  void run() {
+    // We maintain alias db correctness in-place while building up the autodiff
+    // subgraphs, however it is difficult to preserve correctness when
+    // un-inlining autodiff subgraphs. We first recursively construct all
+    // subgraphs and then recursively cleanup & unmerge the small subgraphs
+    buildupSubgraphs();
+    cleanupSubgraphs();
+    // Run CSE globally onceto eliminate duplicates that may have occurred
+    // while inlining subgraphs.
+    EliminateCommonSubexpression(graph_);
+  }
+
+  void cleanupSubgraphs() {
+    auto curNode = *block_->nodes().rbegin();
+    while (curNode != *block_->nodes().rend()) {
+      // Save the previous node, since we might delete `curNode` in next block
+      auto prevNode = curNode->prev();
+      if (curNode->kind() == prim::DifferentiableGraph) {
+        // Inlining nodes may cause some subexpression to come back in the
+        // subgraphs (for example, copying constants in repeatedly will generate
+        // redundant prim::Constants). Run CSE to clean them up.
+        EliminateCommonSubexpression(curNode->g(attr::Subgraph));
```

Comment:
nitpick: should we move CSE into the next check? if we decide to inline it, the final CSE will clean all unmerged graphs up. In other words, we only need to clean a graph up if we aren't inlining it?

Reply:
This is existing logic in the file, I didn't change it. But yea, good point